### ISSUE TITLE
Module load optimisations

### ISSUE THREAD

**Issue by blakeembrey** [jonaskello/tsconfig-paths#1]
It seems like https://github.com/jonaskello/tsconfig-paths/blob/de1ce78b2c24eaefdabbc9c973f49e5b48c55e6b/src/match-path.ts#L55-L57 isn't the best way to do it. It would fall down on other extensions not handled, e.g. `.js`, `.json`, etc. Is it possible to avoid extension checking altogether a return a known path? From the code, it would seem not, so the next best thing - is it possible to use internal module methods to instead resolve the possible paths? This is certainly doable and would be preferred to the custom implementation, as node would then cater to all the known extensions registered for you (e.g. automatically `.ts` and `.tsx` if `ts-node` is running, always `.js` and `.json` built-ins, etc).

**Comment by jonaskello** [comment]
Yes that felt a bit awkward hardcoding the extensions like that so that's definitely something we should improve. I know that typescript itself only allows a [fixed set of extensions](https://github.com/Microsoft/TypeScript/issues/10939) so my thinking when I did that was that we only need to resolve what the typescript compiler itself would resolve. However that does not really apply if you were to execute the compiled JS files without access to the original TS files (you can do this by copying the tsconfig.json to the output and executing it with `node -r tsconfig-paths` main.js). Also it might now apply if you specify `allowJs` to the typescript compiler. And maybe in other cases too?

Skipping the physical file checking altogether would be nice. But since each `paths` pattern can resolve to an array of physical paths I think we need to check if a physical file or dir exists for each physcial path in the array, otherwise we don't know which element to use?

Regarding using internal module methods, that sounds interesting. Are your referring to node's methods or typescript's methods? There is [require().resolve](https://nodejs.org/api/globals.html#globals_require_resolve) in node but I think the logic it uses is quite different (looking in node_modules etc.)

**Comment by jonaskello** [comment]
At the very least I think we need to add checking for `.js` and `.jsx`.

We also need verify the assumption that all JS code that executes originates from TS code and therefore cannot import anything else than the typescript compiler allows. Maybe this assumption is false?

**Comment by Jontem** [comment]
Maybe we could make use of [Module._findPath](https://github.com/nodejs/node/blob/master/lib/module.js#L148). We would then get a lot of functionality already thought of. It will for example handle all the registered extensions for us. 

One thing that looks like a problem is the [tryPackage](https://github.com/nodejs/node/blob/master/lib/module.js#L101) function. It wants to examine `package.json` if the path is a directory which isn't a requirement for paths with typescript(it only requires an index.js).

The question is if paths from tsconfig should cover more extensions that typescript does? I'm not familiar with the allowJs option does it allow you to require more than `.js` files?

**Comment by blakeembrey** [comment]
It seems that theoretically someone could use this package with JavaScript directly (e.g. TypeScript-emitted content) or with loading other files (e.g. CSS, etc). 

> One thing that looks like a problem is the tryPackage function

Ah, I wasn't aware the path rewriting didn't support `package.json`. What about directories with `index.js` in them then? It seems both of those would be support with node module resolution in the TypeScript compiler, but I honestly haven't tested it. If those aren't supported, it's no big deal, just select the extensions manually using `Object.keys(require.extensions)`.

FWIW, you currently have a bug because `existsSync` doesn't check for just files.

> The question is if paths from tsconfig should cover loading of files with extensions that typescript doesn't?

I don't see why it wouldn't?

> Does it allow you to require more than .js files?

You can already require any extension you want today, you just need the matching declaration for TypeScript to allow it.

**Comment by jonaskello** [comment]
I think we need to establish the scope for this package and clarify how it works. In my opinion it should support the cases where `tsc` actually checks in tsconfig's `path` section (rather than being a package for generic path rewriting for other cases). To be more conrete lets look at an example:

`main.ts`
```
import {foo} from "bar";
const foo2 = require("bar");
```

If we compile this example with `tsc`, the `import` is subject to "paths mapping" (as specified under the "path mapping" section in the [handbook](https://www.typescriptlang.org/docs/handbook/module-resolution.html). However the `require` call is not subject to path mapping, instead `tsc` will just emit it as-is without trying to resolve it. You can call `require()` on any file, `.css`, `.json` etc. But as default `tsc` will only allow `import` on:

* Files ending in .ts or .tsx
* Directories that have an index.ts or index.tsx

The way I understand what will happen with the `import` in the example is that first `tsc` will look in the `paths` section of `tsconfig` and try to find a match in the keys. If it finds a match it will check each path in the key's array to see if a physcial file or directory exists. If it exists then it will resolve and end here. If it does not find a match in the keys or a physcial file or dir it will fall back to normal `node_modules` resolution.

This package emualtes the first part, looking in `paths` and resolving to a physical file or dir. If it does not match, it will pass through to node's normal resolving. Therefore I don't think we have to take any special care to make normal `node_modules` resolving work as we already pass non-matches along to node's internal resolving.

There are two cases I can think of, and as mentioned above, where `tsc` would allow other `imports`:

* If you use the `allowJs` compiler option.
* If you add typings for different file endings.

I might have some flaw logic or facts above so please let me know if something is incorrect.

**Comment by jonaskello** [comment]
One problem with supporting only the cases where `tsc` actually do path mapping is that the example above would compile to something like this:

`main.js`
```
const foo = require("bar").foo;
const foo2 = require("bar");
```

So at this point it is not possible to distinguish that the first `require` should be subject to path mapping but the second should not. So we will always try to resolve all `require`which may lead to false positives. Not sure how common false positives would be in practice though.

**Comment by blakeembrey** [comment]
> But as default tsc will only allow import on

This isn't true though, can you import anything as long as you write a definition for it.

**Comment by jonaskello** [comment]
Yes, I think I covered that in the two exception cases mentioned towards the end of my previous post, but maybe a little bit too unclear:

>If you add typings for different file endings.

Just to clarify we are talking about something like this right?:

```
declare module "bar.css";
import foo from "bar.css";
```

**Comment by blakeembrey** [comment]
Gotcha, I just didn't understand that sentence 😄 I'd recommend just using `require.extensions` though - why resolve `.ts` files if once they get loaded they won't work anyway? Or you can just do a union of `Object.keys(require.extensions) + ['.ts', '.tsx']` if you really need to - at least this way all the functionality available in node just works.

Also, you can declare the file without the extension. E.g. `declare './foo'; import './foo'` where `./foo` is a `.css` or `.whatever` file. It will be resolved under node.js if it's in `require.extensions`.

**Comment by jonaskello** [comment]
Agreed, `require.extensions` is probably the best we can do. 

There might be corner cases when we could end up resolving path mappings when `tsc` would not. For example if there is a `.json` file present in a mapped path, then `tsc` would not resolve but instead fall back to normal `node_modules` resolution. But since `.json` exists in require.extensions we will pick it up. So this would be a discrepancy between path mapping in `tsc` and `tsconfig-paths` but in reality I don't think this would happen. 

I've committed the change to use `require.extensions` [here](https://github.com/jonaskello/tsconfig-paths/blob/b14c767735d06e76b31585e7391dcb9a04680f5a/src/match-path.ts#L58-L58). If it looks OK I think we should release it.

**Comment by jonaskello** [comment]
One question that remains is how to handle directories. As it is now, we resolve if a path mapping points to a directory that exists. This assumes that there is an `index` file in that directory with a correct extension but we do not check it. Probably, typescript compilation would fail if the `index` file did not exists so maybe it is unnecessary to check for it?

**Comment by blakeembrey** [comment]
> Probably, typescript compilation would fail if the index file did not exists so maybe it is unnecessary to check for it?

Would it in the edgecase where you're finding the absolute path dynamically and the directory exists in both places but only the second one has an `index` file? It'd be easy enough to run the loop over `x${ext}` and `x/index${ext}`. Might need to figure out how performant some of it is and whether you can do any optimisations without deviating from the TypeScript spec.

**Comment by Jontem** [comment]
Yes i think it's best to do the same resolution logic as typescript does but with with out the searching for type declarations.

As mentioned before typescript starts by looks at path mappings and last it will look in `node_modules`. 

I ran some tests with `tsc --traceResolution` and an import that looks like this `import {barHello} from "lib/bar";`. The resolution order seems to be this:

1. /project/root/src/lib/bar.ts
2. /project/root/src/lib/bar.tsx
3. /project/root/src/lib/bar/package.json
4. /project/root/src/lib/bar/index.ts

Looks like we have to support looking at the main field in package.json. I've verified that typescript uses the main field from package.json and looks if the file exists. Otherwise it moves on.

**Comment by jonaskello** [comment]
I think we only need to support the path mapping part. If it does not match a path mapping we pass through the request to node. So I don't think our package need to concern itself with node's resolution logic at all (such as looking in `package.json` and `node_modules` etc.). In the example we need to take care of step 1 and 2 (but only if they have a path mapping). The other steps will be solved by node as we pass through the request.

**Comment by jonaskello** [comment]
Unless typscript will try node-style resolution for path mappings, but I don't think it will. For example given this path mapping and import:

```
paths: { "foo": [./libs/foo] }
```
  
```
import x from "foo";
```

I think typescript will try to find the file `./libs/foo/index.ts`. But if that fails I think it will just end there. Or will it continue with node style resolution in `./libs/foo/package.json`, `./libs/foo/node_modules/foo` etc.

**Comment by jonaskello** [comment]
@blakeembrey Regarding the case you mentioned, that could certainly happen too, so my assumption that we can rely on the compiler was wrong. And from some experiments @Jontem made there seems to be more going on than the logic we have now. 
I've moved the logic to do the checking to a separate function as it seems it can get more complicated. The function is [here](https://github.com/jonaskello/tsconfig-paths/blob/59b68c32661eae454dd6f648e6d71a45360a42f0/src/match-path.ts#L70-L70)

**Comment by jonaskello** [comment]
My best guess for the correct logic at the moment is:

1. Check for files ending in any of the extensions.
2. Check for a file named `index`ending in any of the extensions.
3. Check for file specified in package.json, `main` property.

I think typescript do step 3 because you should be able to path map to a package. So it does not do full node-style resolution for each mapping, it just allows you to map to a package.

**Comment by jonaskello** [comment]
I have released v1.1.0 which has more advanced logic for resolving. Basically it follows the rules above. Thanks to @Jontem for implementing this!

@blakeembrey Thanks for spotting and driving this issue! I'll close this, but if you spot any problems with the new logic please go ahead and open a new issue.
